diff -r -U5 xfce4-panel-4.8.2/plugins/tasklist/tasklist-widget.c xfce4-panel-4.8.2.ok/plugins/tasklist/tasklist-widget.c
--- xfce4-panel-4.8.2/plugins/tasklist/tasklist-widget.c	2011-02-25 22:58:41.000000000 +0200
+++ xfce4-panel-4.8.2.ok/plugins/tasklist/tasklist-widget.c	2011-03-16 15:27:18.139741010 +0200
@@ -3021,10 +3021,15 @@
         }
     }
 }
 
 
+static void
+xfce_tasklist_group_button_menu_activate(WnckWindow *window)
+{
+  wnck_window_unminimize (window, gtk_get_current_event_time ());
+}
 
 static GtkWidget *
 xfce_tasklist_group_button_menu (XfceTasklistChild *group_child,
                                  gboolean           action_menu_entries)
 {
@@ -3050,10 +3055,12 @@
           gtk_widget_show (mi);
 
           if (action_menu_entries)
             gtk_menu_item_set_submenu (GTK_MENU_ITEM (mi),
                 wnck_action_menu_new (child->window));
+          else 
+             gtk_signal_connect_object(GTK_OBJECT(mi), "activate", G_CALLBACK (xfce_tasklist_group_button_menu_activate), child->window);
         }
     }
 
   if (action_menu_entries)
     {
