--- Info.pm.orig	2011-04-23 12:04:59.000000000 +0300
+++ Info.pm	2011-05-01 09:24:44.000000000 +0300
@@ -264,8 +264,8 @@
 		$@ = "File is empty";
 		return undef;
 	}
-
-	if (ref $file) { # filehandle passed
+	my $ref = ref $file;
+	if ($ref and $ref =~ /Glob/i) { # filehandle passed
 		$fh = $file;
 	} else {
 		if (not open $fh, '+<', $file) {
@@ -414,7 +414,9 @@
 		}
 	}
 
-	if (ref $file) { # filehandle passed
+
+	my $ref = ref $file;
+	if ($ref and $ref =~ /Glob/i) { # filehandle passed
 		$fh = $file;
 	} else {
 		if (not open $fh, '+<', $file) {
@@ -511,7 +513,8 @@
 	}
 
 	# filehandle passed
-	if (ref $file) {
+	my $ref = ref $file;
+	if ($ref and $ref =~ /Glob/i) {
 
 		$fh = $file;
 
@@ -1574,7 +1577,8 @@
 	
 	my $size = -s $file;
 
-	if (ref $file) { # filehandle passed
+	my $ref = ref $file;
+	if ($ref and $ref =~ /Glob/i) { # filehandle passed
 		$fh = $file;
 	} else {
 		if ( !$size ) {
